ddtrace/appsec/_ddwaf.pyx

```diff
@@ -0,0 +1,257 @@
+            idx, val = stack.popleft()
+
+            if isinstance(val, (int, float)):
+                val = str(val)
+
+            if isinstance(val, (six.binary_type, six.text_type)):
+                self._make_string(idx, val)
+
+            elif isinstance(val, Mapping):
+                n = len(val)
+                items_idx = self._reserve_obj(n)
+                self._make_map(idx, items_idx, n)
+                # size of val must not change!! should not happen
+                # while holding the GIL?
+                for j, (k, v) in enumerate(six.iteritems(val)):
+                    if not isinstance(k, (six.binary_type, six.text_type)):
+                        if isinstance(k, (int, float)):
+                            k = str(k)
+                        else:
+                            continue
+                    self._set_parameter(items_idx + j, k)
+                    stack.append((items_idx + j, v))
+
+            elif isinstance(val, Sequence):
+                n = len(val)
+                items_idx = self._reserve_obj(n)
+                self._make_array(idx, items_idx, n)
+                stack.extend([(items_idx + j, val[j]) for j in range(n)])
+
+            i += 1
+
+    def __repr__(self):
+        return "<{0.__class__.__name__} for {0._next_idx} elements>".format(self)
+
+    def __sizeof__(self):
+        return super(_Wrapper, self).__sizeof__() + self._size * sizeof(ddwaf_object)
+
+    def __dealloc__(self):
+        PyMem_Free(self._ptr)
+
+
+cdef class DDWaf(object):
+    """
+    A DDWaf instance performs a matching operation on provided data according
+    to some rules.
+    """
+
+    cdef ddwaf_handle _handle
+    cdef object _rules
+
+    def __init__(self, rules):
+        cdef ddwaf_object* rule_objects
+        self._rules = _Wrapper(rules, max_objects=None)
+        rule_objects = (<_Wrapper?>self._rules)._ptr;
+        self._handle = ddwaf_init(rule_objects, NULL)
+        if <void *> self._handle == NULL:
+            raise ValueError("invalid rules")
+
+    @property
+    def required_data(self):
+        cdef uint32_t size
+        cdef const char* const* ptr
+
+        addresses = []
+        ptr = ddwaf_required_addresses(self._handle, &size)
+        for i in range(size):
+            addresses.append((<bytes> ptr[i]).decode("utf-8"))
+        return addresses
+
+    def run(self, data, timeout_ms=DEFAULT_DDWAF_TIMEOUT_MS):
+        cdef ddwaf_context ctx
+        cdef ddwaf_result result
+
+        ctx = ddwaf_context_init(self._handle, NULL)
+        if <void *> ctx == NULL:
+            raise RuntimeError
+        try:
+            wrapper = _Wrapper(data)
+            ddwaf_run(ctx, (<_Wrapper?>wrapper)._ptr, &result, <uint64_t?> timeout_ms * 1000)
+            if result.data != NULL:
+                return (<bytes> result.data).decode("utf-8")
+        finally:
+            ddwaf_result_free(&result)
```

Comment:
`result` can possibly not be initialized at this point (if `_Wrapper()` raises). Is this ok?

Reply:
Indeed it must not be null

Reply:
Checked with @nizox, this can't be null :)